csi/controller.go

```diff
@@ -115,19 +118,122 @@ func (s *OsdCsiServer) ControllerGetCapabilities(
 // ControllerPublishVolume is a CSI API implements the attachment of a volume
 // on to a node.
 func (s *OsdCsiServer) ControllerPublishVolume(
-	context.Context,
-	*csi.ControllerPublishVolumeRequest,
+	ctx context.Context,
+	req *csi.ControllerPublishVolumeRequest,
 ) (*csi.ControllerPublishVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "This request is not supported")
+
+	logrus.Debugf("Calling Controller Publish Volume in the csi controller")
+	// check if volumeID is present and volume is available on SP
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "VolumeID is mandatory in ControllerPublishVolume request")
+	}
+	clogger.WithContext(ctx).Infof("ControllerPublishVolume request recived. VolumeID: %s", req.GetVolumeId())
+
+	// if nodeID is set, and node is actually present on SP
+	if req.NodeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeID is mandatory in CPVolume request")
+	}
+	// if VolumeCapability is nil, fail.
+	if req.VolumeCapability == nil {
+		return nil, status.Error(codes.InvalidArgument, "VolumeCaps must be present in the NodeSVolReq")
+	}
+	vol, err := s.driverGetVolume(ctx, req.GetVolumeId())
+	if err != nil {
+		if s, ok := status.FromError(err); ok && s.Code() == codes.NotFound {
+			return nil, status.Error(codes.NotFound, "Volume not found")
+		}
+		return nil, err
+	}
+
+	// have the volumeID, submit a new grpc request to px driver to prepare
+	// volume to be published only if the volume propery is winshare.
+	conn, err := s.getRemoteConn(ctx)
+	if err != nil {
+		logrus.Errorf("failed to get remote connection: %v, continuing with local node instead", err)
+		conn, err = s.getConn()
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Unavailable,
+				"Unable to connect to the SDK server: %v", err)
+		}
+	}
+
+	if !vol.Spec.Winshare {
+		// Ignore for the timebeing
+		logrus.Debugf("ControllerPublishVolume: Not a Windows volume, return")
+		return &csi.ControllerPublishVolumeResponse{}, nil
```

Comment:
Maintain the original error which is NotImplemented for unsupported volume types.

Reply:
This is not an error condition, it is no work and an implicit pass if the request to ControllerPublish for a non windows volumes comes through. So current code is correct.

Reply:
Have you tested this? If a CSI driver supports both NodePublish and ControllerPublish, which one is called?

Also, returning a non-error will tell kubernetes that this particular PVC is mounted in the desired container. I'm not sure if that's what we want here.

Reply:
AFAIK, based on the capabilities returned by the driver it will invoke NodePublish and/or ControllerPublish. LNS, I see what you are saying. Yeah may be it makes sense to return no error here since we are now saying we have the ControllerPublish capability.